File: python/problems/leetcode/0337_house_robber_iii/solution.py
"""LeetCode 337: House Robber III."""

from __future__ import annotations

from typing import Optional


class TreeNode:
    def __init__(
        self,
        val: int = 0,
        left: Optional["TreeNode"] = None,
        right: Optional["TreeNode"] = None,
    ) -> None:
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def rob(self, root: Optional[TreeNode]) -> int:
        def dfs(node: Optional[TreeNode]) -> tuple[int, int]:
            if node is None:
                return 0, 0
            ls, lk = dfs(node.left)
            rs, rk = dfs(node.right)
            steal = node.val + lk + rk
            skip = max(ls, lk) + max(rs, rk)
            return steal, skip

        return max(dfs(root))


if __name__ == "__main__":
    root = TreeNode(
        3,
        TreeNode(2, None, TreeNode(3)),
        TreeNode(3, None, TreeNode(1)),
    )
    assert Solution().rob(root) == 7
    root2 = TreeNode(
        3,
        TreeNode(4, TreeNode(1), TreeNode(3)),
        TreeNode(5, None, TreeNode(1)),
    )
    assert Solution().rob(root2) == 9
    assert Solution().rob(None) == 0
    print("0337 OK")

File: cpp/problems/leetcode/0337_house_robber_iii/solution.cpp
// LeetCode 337
#include <alg_std.hpp>
#include <cassert>
using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int v) : val(v), left(nullptr), right(nullptr) {}
    TreeNode(int v, TreeNode* l, TreeNode* r) : val(v), left(l), right(r) {}
};

class Solution {
public:
    int rob(TreeNode* root) {
        auto [s, k] = dfs(root);
        return max(s, k);
    }

private:
    pair<int, int> dfs(TreeNode* node) {
        if (!node) return {0, 0};
        auto [ls, lk] = dfs(node->left);
        auto [rs, rk] = dfs(node->right);
        int steal = node->val + lk + rk;
        int skip = max(ls, lk) + max(rs, rk);
        return {steal, skip};
    }
};

int main() {
    TreeNode n3a(3), n1a(1);
    TreeNode n2(2, nullptr, &n3a), n3b(3, nullptr, &n1a);
    TreeNode root(3, &n2, &n3b);
    assert(Solution().rob(&root) == 7);
    TreeNode l1(1), l3(3), n4(4, &l1, &l3), l1b(1), n5(5, nullptr, &l1b);
    TreeNode root2(3, &n4, &n5);
    assert(Solution().rob(&root2) == 9);
    assert(Solution().rob(nullptr) == 0);
    cout << "0337 OK" << endl;
    return 0;
}
